client/oidc.go

```diff
@@ -23,9 +24,49 @@ type OpenIdProvider interface {
 	PublicKey(ctx context.Context, idt []byte) (crypto.PublicKey, error)
 	VerifyCICHash(ctx context.Context, idt []byte, expectedCICHash string) error
 	VerifyNonGQSig(ctx context.Context, idt []byte, expectedNonce string) error
+	GetIssuer() string
 }
 
+// Interface for interacting with the OP (OpenID Provider)
+type BrowserOpenIdProvider interface {
+	OpenIdProvider
+	HookHTTPSession(h HttpSessionHook)
+}
+type HttpSessionHook func(w http.ResponseWriter, r *http.Request)
+
 func VerifyPKToken(ctx context.Context, pkt *pktoken.PKToken, provider OpenIdProvider) error {
+	return PKTokenVerifer{
+		AllowedProviders: []OpenIdProvider{provider},
+	}.Verify(ctx, pkt)
+}
+
+type PKTokenVerifer struct {
+	AllowedProviders []OpenIdProvider
+	AllowedCosigners []CosignerProvider
+}
+
+func (v PKTokenVerifer) Verify(ctx context.Context, pkt *pktoken.PKToken) error {
+	// If no providers (OPs) are set in the allowlist verify will always fail.
+	if len(v.AllowedProviders) == 0 {
+		return fmt.Errorf("no allowed providers (allowed OPs) set, you must set at least one provider")
+	}
+
+	var pktOp struct {
+		Issuer string `json:"iss"`
+	}
+	if err := json.Unmarshal(pkt.Payload, &pktOp); err != nil {
+		return err
+	}
+	var provider OpenIdProvider
+	for _, allowedOp := range v.AllowedProviders {
+		if allowedOp.GetIssuer() == pktOp.Issuer {
+			provider = allowedOp
+			break
+		}
+	}
+	if provider == nil {
+		return fmt.Errorf("the OP issuer (%s) in the PK Token is not an allowed issuer", pktOp.Issuer)
+	}
```

Comment:
A comment makes sense here. I eventually want to enable the use of the same issuer with different audiences, I'm not sure the best way to do that since audience isn't exposed by the OP (provider) struct and even if it was, we would need to be careful not to try to match it for OPs that use the audience for the CIC.

Probably the ideal solution is a function on the OP interface that allows each OP struct to check if a PK Token "matches" without doing the full round trip to the JWKS endpoint.

Reply:
I created an issue for this and added a comment to document it. As this PR touches too much code, I plan to address this in a future PR

https://github.com/openpubkey/openpubkey/issues/84